test/instrumentation/dropped-span-stats.js

```diff
@@ -0,0 +1,133 @@
+      service: {}
+    })
+    test.ok(!transaction.captureDroppedSpan(span))
+
+    transaction.end()
+    test.end()
+  })
+
+  suite.test('test DroppedSpanStats objects', function (test) {
+    const transaction = agent.startTransaction('trans')
+    for (let i = 0; i < 2; i++) {
+      const span = agent.startSpan('foo', 'baz', 'bar', { exitSpan: true })
+      span.setDestinationContext(destinationContext)
+      span.setOutcome(OUTCOME_SUCCESS)
+      span.end()
+      test.ok(
+        transaction.captureDroppedSpan(span)
+      )
+    }
+
+    for (let i = 0; i < 3; i++) {
+      const span = agent.startSpan('foo', 'baz', 'bar', { exitSpan: true })
+      span.setDestinationContext(destinationContext)
+      span.setOutcome(OUTCOME_FAILURE)
+      span.end()
+      test.ok(
+        transaction.captureDroppedSpan(span)
+      )
+    }
+
+    for (let i = 0; i < 4; i++) {
+      const span = agent.startSpan('foo', 'baz', 'bar', { exitSpan: true })
+      span.setDestinationContext({
+        service: {
+          resource: 'bar'
+        }
+      })
+      span.setOutcome(OUTCOME_SUCCESS)
+      span.end()
+      span._duration = 1000 // override duration so we can test the sum
+      test.ok(
+        transaction.captureDroppedSpan(span)
+      )
+    }
+    transaction.end()
+
+    // three distinct resource/outcome pairs captured
+    test.equals(transaction._droppedSpanStats.statsMap.size, 3)
+
+    const payload = transaction._encode()
+    const stats = payload.dropped_spans_stats
+    test.equals(stats[0].duration.count, 2)
+    test.equals(stats[0].destination_service_resource, 'foo')
+    test.equals(stats[0].outcome, OUTCOME_SUCCESS)
+
+    test.equals(stats[1].duration.count, 3)
+    test.equals(stats[1].destination_service_resource, 'foo')
+    test.equals(stats[1].outcome, OUTCOME_FAILURE)
+
+    test.equals(stats[2].duration.count, 4)
+    test.equals(stats[2].destination_service_resource, 'bar')
+    test.equals(stats[2].duration.sum.us, 4000000)
+    test.equals(stats[2].outcome, OUTCOME_SUCCESS)
+
+    test.end()
+  })
+
+  suite.test('test DroppedSpanStats max items', function (test) {
+    const transaction = agent.startTransaction('trans')
+    for (let i = 0; i < 128; i++) {
+      const destinationContext = {
+        service: {
+          resource: 'foo' + i
+        }
+      }
+      const span = agent.startSpan('foo', 'baz', 'bar', { exitSpan: true })
+      span.setDestinationContext(destinationContext)
+      span.setOutcome(OUTCOME_FAILURE)
+      span.end()
+      test.ok(transaction.captureDroppedSpan(span))
+    }
+
+    // one too many
+    const span = agent.startSpan('foo', 'baz', 'bar', { exitSpan: true })
+    span.setDestinationContext(destinationContext)
+    span.setOutcome(OUTCOME_FAILURE)
+    span.end()
+
+    test.ok(!transaction.captureDroppedSpan(span))
+    transaction.end()
+    test.equals(transaction._droppedSpanStats.statsMap.size, 128)
+    test.end()
+  })
+
+  suite.end()
```

Comment:
It would be good to have a test case for `transaction.captureDroppedSpan(span)` automatically being called. So, a case where `transactionMaxSpans` is hit and the resulting serialized transaction includes the expected `dropped_spans_stats: [...]`.   A quick place to do this might be in this existing transaction.test.js case that exercises `transactionMaxSpans`: https://github.com/elastic/apm-agent-nodejs/blob/main/test/instrumentation/transaction.test.js#L493-L529

Reply:
8437013896f655dbbe973c3fddbcfb54b3f29a10

Reply:
I don't see where that commit adds a test case for this.

Reply:
New test code is here: https://github.com/elastic/apm-agent-nodejs/commit/8437013896f655dbbe973c3fddbcfb54b3f29a10#diff-115ec49fff221477451b17e1fca2a5208448489ccbb84c4eb7b53c6c4016f3caR536

We ensure that the serialized payload contains data for dropped_spans_stats.

```
    t.equals(payload.dropped_spans_stats.length, 1)
```

If this isn't what you wanted to see please let me know what would make this an acceptable test to you.

Reply:
Thanks. I had been trying to find it in the [dropped-span-stats.test.js](https://github.com/elastic/apm-agent-nodejs/commit/8437013896f655dbbe973c3fddbcfb54b3f29a10#diff-bb4011f7f490e34ded1fbef4a570aedc94fd1a3c2bb563791678f413ca0c4927) file part of the diff.